libbpf-tools/trace_helpers.c

```diff
@@ -150,3 +150,70 @@ const struct ksym *ksyms__get_symbol(const struct ksyms *ksyms,
 
 	return NULL;
 }
+
+#define min(x, y) ({				 \
+	typeof(x) _min1 = (x);			 \
+	typeof(y) _min2 = (y);			 \
+	(void) (&_min1 == &_min2);		 \
+	_min1 < _min2 ? _min1 : _min2; })
+
+static void print_stars(unsigned int val, unsigned int val_max, int width)
+{
+	int num_stars, num_spaces, i;
+	bool need_plus;
+
+	num_stars = min(val, val_max) * width / val_max;
+	num_spaces = width - num_stars;
+	need_plus = val > val_max;
+
+	for (i = 0; i < num_stars; i++)
+		printf("*");
+	for (i = 0; i < num_spaces; i++)
+		printf(" ");
+	if (need_plus)
+		printf("+");
+}
+
+void print_log2_hist(unsigned int *vals, int vals_size, char *val_type)
+{
+	int stars_max = 40, idx_max = -1;
+	unsigned int val, val_max = 0;
+	unsigned long long low, high;
+	int stars, i;
+
+	for (i = 0; i < vals_size; i++) {
+		val = vals[i];
+		if (val > 0)
+			idx_max = i;
+		if (val > val_max)
+			val_max = val;
+	}
+
+	if (idx_max > 0) {
+		if (idx_max <= 32)
+			printf("     %-19s : count    distribution\n",
+				val_type);
+		else
+			printf("               %-29s : count     "
+				"distribution\n", val_type);
+	}
+
+	if (idx_max <= 32)
+		stars = stars_max;
+	else
+		stars = stars_max / 2;
+
+	for (i = 1; i < idx_max + 1; i++) {
+		low = (1ULL << i) >> 1;
+		high = (1ULL << i) - 1;
+		if (low == high)
+			low -= 1;
+		val = vals[i];
```

Comment:
this will ignore slot #0, that doesn't sound right. 

I also think that your histogram logic is wrong and log2l is wrong as well. I printed out what log2 and log2l return  for i=0..32, see for yourself:
```
i=0 log2(0)=0 log2l(0)=1
i=1 log2(1)=0 log2l(1)=1
i=2 log2(2)=1 log2l(2)=2
i=3 log2(3)=1 log2l(3)=2
i=4 log2(4)=2 log2l(4)=3
i=5 log2(5)=2 log2l(5)=3
i=6 log2(6)=2 log2l(6)=3
i=7 log2(7)=2 log2l(7)=3
i=8 log2(8)=3 log2l(8)=4
i=9 log2(9)=3 log2l(9)=4
i=10 log2(10)=3 log2l(10)=4
i=11 log2(11)=3 log2l(11)=4
i=12 log2(12)=3 log2l(12)=4
i=13 log2(13)=3 log2l(13)=4
i=14 log2(14)=3 log2l(14)=4
i=15 log2(15)=3 log2l(15)=4
i=16 log2(16)=4 log2l(16)=5
i=17 log2(17)=4 log2l(17)=5
i=18 log2(18)=4 log2l(18)=5
i=19 log2(19)=4 log2l(19)=5
i=20 log2(20)=4 log2l(20)=5
i=21 log2(21)=4 log2l(21)=5
i=22 log2(22)=4 log2l(22)=5
i=23 log2(23)=4 log2l(23)=5
i=24 log2(24)=4 log2l(24)=5
i=25 log2(25)=4 log2l(25)=5
i=26 log2(26)=4 log2l(26)=5
i=27 log2(27)=4 log2l(27)=5
i=28 log2(28)=4 log2l(28)=5
i=29 log2(29)=4 log2l(29)=5
i=30 log2(30)=4 log2l(30)=5
i=31 log2(31)=4 log2l(31)=5
i=32 log2(32)=5 log2l(32)=6
```
log2l is always +1 from log2, this is completely wrong, I don't understand why log2l has +1 in there. I think log2() results make sense, though, given log2(X) + 1 should give minimum number of bits necessary to represent a given X.

So with that, values of 0 and 1 go into slot #0, while you ignore them and report values in slot #1 as [0, 1], while they should be [2..3], and so on.

So please fix that and double check that everything is accurate. It's hard to trust a tool that doesn't do math correctly :)

Reply:
Thanks a lot, I'll do more check at the weekends to make sure everything is accurate, and thanks again for your patient guidance.

Reply:
> log2l is always +1 from log2, this is completely wrong, I don't understand why log2l has +1 in there. I think log2() results make sense, though, given log2(X) + 1 should give minimum number of bits necessary to represent a given X.

Sorry for this! I migrate hist and log2 from bcc version directly. I thought it's ok, so I didn't think about it carefully.  I found https://github.com/iovisor/bcc/issues/1649 today, it looks like a history's legac. 

I've fixed with zero slot as beginning. Thanks.